[Cocoa] Adopt CASpatialAudioExperience

jernoble opened the pull request:
#### 2f5b4a7bc46ee9003355321d82e59eb11fc68a80
<pre>
[Cocoa] Adopt CASpatialAudioExperience
<a href="https://rdar.apple.com/145320517">rdar://145320517</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=288240">https://bugs.webkit.org/show_bug.cgi?id=288240</a>

Reviewed by Andy Estes.

Add support for CASpatialAudioExperience as a replacement for spatial tracking labels.

CASpatialAudioExperiences require a few pieces of information: whether the view is
currently visible (i.e., the web view isn&apos;t currently in a background tab that&apos;s not
rooted in the view heirarchy), whether the audio is automatically associated with
a visible layer (as is the case with AVPlayerLayer and AVSampleBufferRenderSynchronizer).
Currently the behavior is re-implemented across multiple different MediaPlayer subtypes,
so provide a helper function for creating these experience objects that can be
re-used.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::sceneIdentifierDidChange):
(WebCore::AudioContext::sceneIdentifier const):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/Modules/webaudio/AudioDestinationNode.h:
(WebCore::AudioDestinationNode::setSceneIdentifier):
* Source/WebCore/Modules/webaudio/BaseAudioContext.h:
(WebCore::BaseAudioContext::sceneIdentifier const):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
(WebCore::DefaultAudioDestinationNode::createDestination):
(WebCore::DefaultAudioDestinationNode::setSceneIdentifier):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/AudioToolboxCoreSPI.h: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::sceneIdentifierDidChange):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::sceneIdentifierDidChange):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/MediaProducer.h:
(WebCore::MediaProducer::sceneIdentifierDidChange):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::sceneIdentifier const):
(WebCore::Page::setSceneIdentifier):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::setSceneIdentifier):
* Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.h: Added.
* Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.mm: Added.
(WebCore::createExperienceWithOptions):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::setSceneIdentifier):
(WebCore::MediaPlayer::setPrefersSpatialAudioExperience):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::prefersSpatialAudioExperienceChanged):
(WebCore::MediaPlayerPrivateInterface::sceneIdentifierDidChange):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::sceneIdentifierDidChange):
(WebCore::MediaPlayerPrivateAVFoundationObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::sceneIdentifierDidChange):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::updateSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateWebM::sceneIdentifierDidChange):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::setPrefersSpatialAudioExperience):
(WebCore::VideoPresentationInterfaceIOS::prefersSpatialAudioExperience const):
* Source/WebCore/platform/network/ios/WebCoreURLResponseIOS.mm:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setSceneIdentifier):
(WebKit::RemoteMediaPlayerProxy::setPrefersSpatialAudioExperience):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::captionsLayer):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createModelAndInterface):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::prefersSpatialAudioExperienceChanged):
(WebKit::MediaPlayerPrivateRemote::sceneIdentifierDidChange):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/291033@main">https://commits.webkit.org/291033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7301cc238d3c7bdd9f92a19f2241db8900731b69

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91284 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10819 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/314 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96267 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42006 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93334 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19136 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/96267 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27632 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82681 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/96267 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8310 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/262 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41149 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84094 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78631 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/272 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98258 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90040 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18952 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13527 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79131 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18715 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78333 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19476 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/23263 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/197 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11628 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18458 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23758 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112618 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18177 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32678 "Found 10 new JSC stress test failures: microbenchmarks/memcpy-wasm-small.js.default, microbenchmarks/memcpy-wasm-small.js.dfg-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.default-wasm, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-tag-arg.js.wasm-eager (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21638 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/PAL/pal/spi/cocoa/AudioToolboxCoreSPI.h:
```diff
@@ -0,0 +1,56 @@
+    CAAudioTetherTypeLayer,
+    CAAudioTetherTypeEntity,
+};
+
```
```suggestion
NS_ASSUME_NONNULL_BEGIN
```

aestes commented on Source/WebCore/PAL/pal/spi/cocoa/AudioToolboxCoreSPI.h:
```diff
@@ -0,0 +1,56 @@
+
+@interface CAAudioTetherAnchoringStrategy : CAAnchoringStrategy
+- (instancetype)initWithAudioTether:(CAAudioTether*)audioTether;
+@end
```
```suggestion
@end

NS_ASSUME_NONNULL_END
```

aestes commented on Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.mm:
```diff
@@ -0,0 +1,60 @@
+        RetainPtr uuid = adoptNS([[NSUUID alloc] initWithUUIDString:options.spatialTrackingLabel]);
+        RetainPtr tether = adoptNS([[CAAudioTether alloc] initWithType:CAAudioTetherTypeLayer identifier:uuid.get() pid:0]);
+        RetainPtr anchoringStrategy = adoptNS([[CAAudioTetherAnchoringStrategy alloc] initWithAudioTether:tether.get()]);
+        return adoptNS([[CAHeadTrackedSpatialAudio alloc] initWithSoundStageSize:CASoundStageSizeAutomatic anchoringStrategy:anchoringStrategy.get()]);
```
How does this sound stage size interact with the one we set on `AVAudioSession`? Do we want to pass in a sound stage size via `SpatialAudioExperienceOptions` rather than hard-coding `CASoundStageSizeAutomatic`?

jernoble replied:
Yes; lets do that in the next patch.

aestes commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -314,15 +314,17 @@ - (instancetype)_commonInitializationWithProcessPool:(WebKit::WebProcessPool&)pr
-    [_spatialTrackingView setFrame:CGRectMake(CGRectGetMidX(self.bounds), CGRectGetMidY(self.bounds), 0, 0)];
-    [_spatialTrackingView setUserInteractionEnabled:NO];
-    [self addSubview:_spatialTrackingView.get()];
+#if HAVE(SPATIAL_TRACKING_LABEL) && HAVE(SPATIAL_AUDIO_EXPERIENCE)
```
Isn't this wrong in a configuration where `HAVE(SPATIAL_TRACKING_LABEL)` is true but `HAVE(SPATIAL_AUDIO_EXPERIENCE)` is false (i.e. we won't set up a `_spatialTrackingView` in that case)?

jernoble committed:
[Cocoa] Adopt CASpatialAudioExperience
rdar://145320517
https://bugs.webkit.org/show_bug.cgi?id=288240

Reviewed by Andy Estes.

Add support for CASpatialAudioExperience as a replacement for spatial tracking labels.

CASpatialAudioExperiences require a few pieces of information: whether the view is
currently visible (i.e., the web view isn't currently in a background tab that's not
rooted in the view heirarchy), whether the audio is automatically associated with
a visible layer (as is the case with AVPlayerLayer and AVSampleBufferRenderSynchronizer).
Currently the behavior is re-implemented across multiple different MediaPlayer subtypes,
so provide a helper function for creating these experience objects that can be
re-used.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::sceneIdentifierDidChange):
(WebCore::AudioContext::sceneIdentifier const):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/Modules/webaudio/AudioDestinationNode.h:
(WebCore::AudioDestinationNode::setSceneIdentifier):
* Source/WebCore/Modules/webaudio/BaseAudioContext.h:
(WebCore::BaseAudioContext::sceneIdentifier const):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
(WebCore::DefaultAudioDestinationNode::createDestination):
(WebCore::DefaultAudioDestinationNode::setSceneIdentifier):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/AudioToolboxCoreSPI.h: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::sceneIdentifierDidChange):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::sceneIdentifierDidChange):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/MediaProducer.h:
(WebCore::MediaProducer::sceneIdentifierDidChange):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::sceneIdentifier const):
(WebCore::Page::setSceneIdentifier):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::setSceneIdentifier):
* Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.h: Added.
* Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.mm: Added.
(WebCore::createExperienceWithOptions):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::setSceneIdentifier):
(WebCore::MediaPlayer::setPrefersSpatialAudioExperience):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::prefersSpatialAudioExperienceChanged):
(WebCore::MediaPlayerPrivateInterface::sceneIdentifierDidChange):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::sceneIdentifierDidChange):
(WebCore::MediaPlayerPrivateAVFoundationObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::sceneIdentifierDidChange):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::updateSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateWebM::sceneIdentifierDidChange):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
(WebCore::VideoPresentationInterfaceIOS::setPrefersSpatialAudioExperience):
(WebCore::VideoPresentationInterfaceIOS::prefersSpatialAudioExperience const):
* Source/WebCore/platform/network/ios/WebCoreURLResponseIOS.mm:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setSceneIdentifier):
(WebKit::RemoteMediaPlayerProxy::setPrefersSpatialAudioExperience):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::captionsLayer):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createModelAndInterface):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::prefersSpatialAudioExperienceChanged):
(WebKit::MediaPlayerPrivateRemote::sceneIdentifierDidChange):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: https://commits.webkit.org/291033@main
